crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2576,7 +2609,53 @@ impl<'db> TypeInferenceBuilder<'db> {
                         Type::Unknown
                     })
             }
-            _ => Type::Todo,
+            (value_ty, slice_ty) => {
+                // Resolve the value to its class.
+                let value_meta_ty = value_ty.to_meta_type(self.db);
+
+                // If the class defines `__getitem__`, return its return type.
+                //
+                // See: https://docs.python.org/3/reference/datamodel.html#class-getitem-versus-getitem
+                let dunder_getitem_method = value_meta_ty.member(self.db, "__getitem__");
+                if !dunder_getitem_method.is_unbound() {
+                    // TODO validate that the `__getitem__` argument adheres to the signature
+                    let CallOutcome::Callable { return_ty } =
+                        dunder_getitem_method.call(self.db, &[slice_ty])
+                    else {
+                        self.dunder_not_callable_diagnostic(
+                            (&**value).into(),
+                            value_ty,
+                            "__getitem__",
+                        );
+                        return Type::Unknown;
+                    };
+                    return return_ty;
+                }
+
+                // Otherwise, if the value is itself a class and defines `__class_getitem__`,
+                // return its return type.
+                if matches!(value_ty, Type::Class(_)) {
```

Comment:
I think we should add `Type::is_class` (true if it's a `Type::Class` or a union of them) instead of directly matching on the `Type` variant this way. The reason is because we should also support `__class_getitem__` on a union of classes. `Type::member` and `Type::call` both already support unions, so it should "just work" if we don't prevent it here.

And lets also add a test for this case.

Reply:
When we add support for `type[T]`, we'll need to remember to modify `Type::is_class()` so that it returns `true` for `type[T]` as well. `x.__class_getitem__` should not be an error if the type of `x` is `type[dict]` (since all subclasses of `dict` have `__class_getitem__` methods as well as `dict`).

Nothing to do for now, but a TODO comment in `Type::is_class` might be good

Reply:
So the test case here would be "The class is a union" rather than "`__class_getitem` is a union", right?

Reply:
Yeah -- in part because the former will necessarily imply the latter (when you get a member of a union, you get the union of the members of all the constituent types of the union)